### PR TITLE
Dev/add ipc support

### PR DESCRIPTION
**Pull request by kfieldho**

### CODE REVIEW

File: arrows/serialize/json/tests/test_serialize.cxx [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+  auto mes = bbox_ser.serialize( sp );
+
+
+  // check element names
+  const auto& names = bbox_ser.element_names();
+
+  EXPECT_EQ( names.size(), 1 );
+
+  // std::cout << "Serialized bbox: \"" << *mes << "\"\n";
+  // std::cout << "List of element names: " << kwiver::vital::join( names, ", " ) << std::endl;
```
Comment (dstoup) [inline arrows/serialize/json/tests/test_serialize.cxx]: Remove dead code.

File: arrows/serialize/json/tests/test_serialize.cxx [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+  auto dser = bbox_ser.deserialize( mes );
+  kwiver::vital::bounding_box_d bbox_dser =
+    kwiver::vital::any_cast< kwiver::vital::bounding_box_d >( dser[ kwiver::vital::algo::data_serializer::DEFAULT_ELEMENT_NAME ] );
+
+  /*
+  std::cout << "bbox_dser { " << bbox_dser.min_x() << ", "
+            << bbox_dser.min_y() << ", "
+            << bbox_dser.max_x() << ", "
+            << bbox_dser.max_y() << "}\n";
+  */
```
Comment (dstoup) [inline arrows/serialize/json/tests/test_serialize.cxx]: Remove dead code.

File: arrows/serialize/protobuf/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+set( sources
+  bounding_box.cxx
+  detected_object_type.cxx
+  detected_object.cxx
+  detected_object_set.cxx
+  )
+
+link_directories(${PROTOBUF_LIBRARY})
+include_directories(${PROTOBUF_INCLUDE_DIR})
+add_compile_options( -include "${CMAKE_BINARY_DIR}/vital/types/protobuf/vital_types_protobuf_export.h" )
```
Comment (dstoup) [inline arrows/serialize/protobuf/CMakeLists.txt]: I am not 100% sure what this command is providing, but I see it didn't appear in the json CMakeLists.txt. Should it be there?

File: vital/algo/data_serializer.h [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+ * objects for inter-process communications.
+ */
+class VITAL_ALGO_EXPORT data_serializer
+  : public kwiver::vital::algorithm_def< data_serializer >
+{
+public:
+  using serialize_param_t = std::map< std::string, vital::any >;
+  using deserialize_result_t = std::map< std::string, vital::any >;
+
+/// Return the name of this algorithm
```
Comment (Purg) [inline vital/algo/data_serializer.h]: Not spaced evenly with function.

File: arrows/serialize/protobuf/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+kwiver_add_library( kwiver_serialize_protobuf
+  ${headers_public}
+  ${sources}
+  )
+
+target_link_libraries( kwiver_serialize_protobuf
+  PUBLIC               vital
+                       vital_vpm
+  PRIVATE              vital_algo
+                       vital_exceptions
```
Comment (dstoup) [inline arrows/serialize/protobuf/CMakeLists.txt]: I think vital_algo is in an included header, which might require it to be PUBLIC. Hard to follow the paper trail though, could be wrong.

File: arrows/serialize/json/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+kwiver_add_library( kwiver_serialize_json
+  ${headers_public}
+  ${sources}
+  )
+
+target_link_libraries( kwiver_serialize_json
+  PUBLIC               vital
+                       vital_vpm
+  PRIVATE              vital_algo
```
Comment (dstoup) [inline arrows/serialize/json/CMakeLists.txt]: I think vital_algo is in an included header, which might require it to be PUBLIC. Hard to follow the paper trail though, could be wrong.

File: arrows/serialize/protobuf/CMakeLists.txt (hunk 5 above)
Comment (dstoup) [reply arrows/serialize/protobuf/CMakeLists.txt]: Similarly, it looks like vital_vpm's headers aren't included in this libraries heads, which means it probably doesn't need to be a PUBLIC link. The same might apply for the json serialization as well.

File: arrows/serialize/protobuf/tests/test_serialize.cxx [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    auto o_it = dot->begin();
+    auto d_it = dot_dser->begin();
+
+    for (size_t i = 0; i < dot->size(); ++i )
+    {
+      EXPECT_EQ( *(o_it->first), *(d_it->first) );
+      EXPECT_EQ( o_it->second, d_it->second );
+    }
+  }
+  */
```
Comment (dstoup) [inline arrows/serialize/protobuf/tests/test_serialize.cxx]: Is the above dead code needed?

File: sprokit/processes/transport/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+kwiver_private_header_group( ${private_headers} )
+
+include_directories( ${CMAKE_CURRENT_BINARY_DIR} )
+
+kwiver_add_plugin(processes_transport
+  SUBDIR      ${kwiver_plugin_process_subdir}
+  SOURCES     ${sources}
+              ${private_headers}
+  PRIVATE     sprokit_pipeline vital_vpm ${ZeroMQ_LIBRARIES}
```
Comment (dstoup) [inline sprokit/processes/transport/CMakeLists.txt]: zmq_transport_send_process.h includes zmq.hpp which means ${ZeroMQ_LIBRARIES} should be a PUBLIC link.

File: CMake/kwiver-depends-ZeroMQ.cmake [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Optionally find and configure CUDA dependency
```
[the remark this answers is not in the document]
Comment (linus-sherrill) [reply CMake/kwiver-depends-ZeroMQ.cmake]: Done

File: arrows/core/detected_object_set_input_simulator.cxx [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+detected_object_set_input_simulator::
+open( std::string const& filename )
+{
+}
+
+
+// ------------------------------------------------------------------
+bool
+detected_object_set_input_simulator::
+read_set( kwiver::vital::detected_object_set_sptr & detected_set, std::string& image_name )
```
[the remark this answers is not in the document]
Comment (linus-sherrill) [reply arrows/core/detected_object_set_input_simulator.cxx]: All parameters are output parameters. Values are returned through the references.

File: arrows/serialize/json/detected_object.cxx [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+
+detected_object::
+~detected_object()
+{ }
+
+// ----------------------------------------------------------------------------
+std::shared_ptr< std::string >
+detected_object::
+serialize( const serialize_param_t elements )
```
[the remark this answers is not in the document]
Comment (linus-sherrill) [reply arrows/serialize/json/detected_object.cxx]: Yes it does.

File: arrows/serialize/protobuf/CMakeLists.txt (hunk 3 above)
Comment (linus-sherrill) [reply arrows/serialize/protobuf/CMakeLists.txt]: # Need to add this include file to the compile because of a defficiency in the protobuf
# compiler. The compiler has the option to insert an export macro into class definitions
# but has no way of defining that macro. as of libprotoc 2.6.1 this is the accepted way
# of defining the export symbols.

File: arrows/serialize/protobuf/CMakeLists.txt (hunk 5 above)
Comment (linus-sherrill) [reply arrows/serialize/protobuf/CMakeLists.txt]: Changed visibility to be more like other plugin libraries

File: sprokit/processes/transport/CMakeLists.txt (hunk 8 above)
Comment (linus-sherrill) [reply sprokit/processes/transport/CMakeLists.txt]: Nothing links against processes, so no external visibility is needed.

File: vital/algo/data_serializer.h (hunk 4 above)
Comment (linus-sherrill) [reply vital/algo/data_serializer.h]: Done

**Comment by dstoup** [comment]
FYI, Here are the Windows build errors.


"C:\Jenkins\workspace\KwiverWindowsPR\build\ALL_BUILD.vcxproj" (default target) (1) ->
"C:\Jenkins\workspace\KwiverWindowsPR\build\vital\algo\algo_explorer_plugin.vcxproj" (default target) (3) ->
"C:\Jenkins\workspace\KwiverWindowsPR\build\vital\tools\explorer_plugin.vcxproj" (default target) (4) ->
"C:\Jenkins\workspace\KwiverWindowsPR\build\vital\vital.vcxproj" (default target) (6) ->
"C:\Jenkins\workspace\KwiverWindowsPR\build\vital\config\vital_config.vcxproj" (default target) (7) ->
"C:\Jenkins\workspace\KwiverWindowsPR\build\vital\util\vital_util.vcxproj" (default target) (10) ->
(ClCompile target) -> 
  c:\jenkins\workspace\kwiverwindowspr\vital\util\hex_dump.cxx(54): error C2039: 'isprint': is not a member of 'std' [C:\Jenkins\workspace\KwiverWindowsPR\build\vital\util\vital_util.vcxproj]
  c:\jenkins\workspace\kwiverwindowspr\vital\util\hex_dump.cxx(62): error C2039: 'min': is not a member of 'std' [C:\Jenkins\workspace\KwiverWindowsPR\build\vital\util\vital_util.vcxproj]
  c:\jenkins\workspace\kwiverwindowspr\vital\util\hex_dump.cxx(62): error C3861: 'min': identifier not found [C:\Jenkins\workspace\KwiverWindowsPR\build\vital\util\vital_util.vcxproj]

**Comment by dstoup** [comment]
LGTM
